File: 03/ICPC2020-Prelim/B/ans-uf.cpp
#include <bits/stdc++.h>
using namespace std;

struct UnionFind {
  vector<int> par;
  UnionFind(int n): par(n, -1) {}

  // 2頂点を結合する
  int unite(int x, int y) {
    int a = root(x), b = root(y);
    if (a == b) return a;
    if (-par[a] < -par[b]) swap(a, b);
    par[a] += par[b];
    par[b] = a;
    return a;
  }

  // 2頂点は連結か
  bool same(int x, int y) {
    return root(x) == root(y);
  }

  // 頂点xの木の根を返す
  int root(int x) {
    if (par[x] < 0) return x;
    return par[x] = root(par[x]);
  }

  // 頂点xが属する木のサイズを返す
  int size(int x) {
    return -par[root(x)];
  }
};

int main() {
  while (true) {
    // 入力
    int m, n, p;
    cin >> m >> n >> p;
    if (m + n + p == 0) break;

    // m頂点からなるUnionFind木を作る
    UnionFind uf(m);

    for (int i = 0; i < n; i++) {
      int a, b;
      cin >> a >> b;

      // もしも最初の感染者pとaが同じグループ→aも感染疑いがある
      // あるいはpとbが同じグループ→bも感染疑いがある
      if (uf.same(p - 1, a - 1) || uf.same(p - 1, b - 1)) {
        // aとbを結合する
        // →pとaとbは同じグループになる
        // →aとbは2人とも感染疑いがある
        uf.unite(a - 1, b - 1);
      }
    }

    // pと同じグループの人数 = 感染疑いのある人数 を出力
    cout << uf.size(p - 1) << endl;
  }
  return 0;
}

File: 03/ICPC2020-Prelim/B/ans-uf.py
class UnionFind:
    def __init__(self, n):
        self.par = [-1 for i in range(n)]

    # 2頂点を結合する
    def unite(self, x, y):
        a = self.root(x)
        b = self.root(y)
        if a != b:
            if -self.par[a] < -self.par[b]:
                t = a
                a = b
                b = t
            self.par[a] += self.par[b]
            self.par[b] = a
        return a

    # 2頂点は連結か
    def same(self, x, y):
        return self.root(x) == self.root(y)

    # 頂点xの木の根を返す
    def root(self, x):
        if self.par[x] < 0:
            return x
        else:
            self.par[x] = self.root(self.par[x])
            return self.par[x]

    # 頂点xが属する木のサイズを返す
    def size(self, x):
        return -self.par[self.root(x)]


while True:
    # 入力
    m, n, p = map(int, input().split())
    if m + n + p == 0:
        break

    # m頂点からなるUnionFind木を作る
    uf = UnionFind(m)

    for i in range(n):
        a, b = map(int, input().split())

        # もしも最初の感染者pとaが同じグループ→aも感染疑いがある
        # あるいはpとbが同じグループ→bも感染疑いがある
        if uf.same(p - 1,  a - 1) or uf.same(p - 1,   b - 1):
            # aとbを結合する
            # →pとaとbは同じグループになる
            # →aとbは2人とも感染疑いがある
            uf.unite(a - 1, b - 1)

    # pと同じグループの人数 = 感染疑いのある人数 を出力
    print(uf.size(p - 1))
